aeron-driver/src/main/c/aeron_driver_conductor.c

```diff
@@ -315,6 +322,88 @@ bool aeron_client_has_reached_end_of_life(aeron_driver_conductor_t *conductor, a
     return client->reached_end_of_life;
 }
 
+int32_t aeron_driver_conductor_next_session_id(aeron_driver_conductor_t* conductor)
+{
+    const int32_t low = conductor->publication_reserved_session_id_low;
+    const int32_t high = conductor->publication_reserved_session_id_high;
+
+    return
+        low <= conductor->next_session_id && conductor->next_session_id <= high
+            ? high + 1
+            : conductor->next_session_id;
+}
+
+int32_t aeron_driver_conductor_update_next_session_id(aeron_driver_conductor_t* conductor, int32_t session_id)
+{
+    conductor->next_session_id = session_id + 1;
+    return session_id;
+}
+
+static void aeron_driver_conductor_track_session_id_offsets(
+    aeron_driver_conductor_t *conductor,
+    aeron_bit_set_t *session_id_offsets,
+    int32_t publication_session_id)
+{
+    const int32_t session_id_offset = publication_session_id - aeron_driver_conductor_next_session_id(conductor);
+    if (0 <= session_id_offset && (size_t) session_id_offset < session_id_offsets->bit_count)
+    {
+        aeron_bit_set_set(session_id_offsets, (size_t) session_id_offset, true);
+    }
+}
+
+static int aeron_driver_conductor_speculate_next_session_id(
+    aeron_driver_conductor_t *conductor,
+    aeron_bit_set_t *session_id_offsets,
+    int32_t *session_id)
+{
+    size_t index = 0;
+
+    if (0 > aeron_bit_set_find_first(session_id_offsets, false, &index))
+    {
+        return -1;
+    }
+
+    *session_id = aeron_driver_conductor_next_session_id(conductor) + (int) index;
+    return 0;
+}
+
+int aeron_confirm_publication_match(
+    const char *uri, int32_t stream_id, aeron_uri_publication_params_t *params, int32_t existing_session_id,
+    aeron_logbuffer_metadata_t *logbuffer_metadata)
+{
+    if (params->has_session_id && params->session_id != existing_session_id)
+    {
+        aeron_set_err(
+            EINVAL,
+            "Existing session-id [%d] doesn't match specified channel: %s, stream-id: %" PRId32,
+            existing_session_id, uri, stream_id);
```

Comment:
channel uri from client may not be null terminated, length must also be considered.

Reply:
Remove uri from the method and made the error strings match those in the Java driver.